File: backend/src/main/java/de/codeministry/leadgen/config/SourceDetail.java
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * Copyright 2026 Marcello Muscara (codeministry)
 *
 * Licensed under the Apache License, Version 2.0. You may obtain a copy of the
 * License at http://www.apache.org/licenses/LICENSE-2.0
 */
package de.codeministry.leadgen.config;

import java.util.List;

/**
 * One source, opened: what defines it and what it has been doing.
 *
 * <p>One record and one request, because one disclosure is one click. Split in two, the panel
 * would carry two loading states and two failure modes for a single decision.
 *
 * <p>Nothing here is on {@link SourceSummary}. That list is refetched whenever a run finishes
 * and whenever the tab comes back to the front, and configuration text has no business on that
 * path for a panel most readers never open.
 *
 * @param id         the source, as {@code sources.yaml} names it
 * @param kind       whatever {@code type} that block declares
 * @param enabled    whether a run would read it
 * @param file       which file defines it, and from which of the two layers
 * @param block      that file's own lines for this source, secrets masked
 * @param connection the block of the connection it names, when it names one. An {@code imap}
 *                   source is half-defined by a block somewhere else in the file, and a reader
 *                   sent to go and find it will find the credentials rather than the settings.
 * @param runs       what it did, newest first, at most what was asked for
 * @param trend      when the numbers last moved, so the panel can say it in a sentence
 */
public record SourceDetail(
    String id,
    String kind,
    boolean enabled,
    ConfigFile file,
    YamlBlock block,
    YamlBlock connection,
    List<SourceRun> runs,
    SourceTrend trend) {

    /**
     * @param name   the file's own name, which is the same in both layers
     * @param layer  {@code default} for the copy inside the jar, {@code config-dir} for one
     *               outside it. One value for the whole file: the two layers override each
     *               other file by file and never key by key, which is exactly what the column
     *               this replaced was claiming to know per row.
     * @param origin where it was read from. A path on somebody's machine, so it belongs in a
     *               panel one deliberate click away and never in a header line that is in
     *               every screenshot.
     */
    public record ConfigFile(String name, String layer, String origin) {
    }
}

File: frontend/src/app/core/model/source-detail.ts
import {ConfigLayer, SourceKind} from './source-summary';

/**
 * A piece of a configuration file, as it is written in it.
 *
 * The line range is what makes the excerpt checkable: a reader can open the file and find
 * exactly these lines, rather than trusting that this is what is in it.
 */
export interface YamlBlock {
  readonly text: string;
  readonly firstLine: number;
  readonly lastLine: number;
}

/**
 * What one source did on one run.
 *
 * A day and not an instant: the time of day a mailbox is read is the operator's working
 * hours, it is on a screen whose pictures get published, and nothing here needs it.
 */
export interface SourceRun {
  readonly ranOn: string;
  readonly documents: number;
  readonly extracted: number;
  /** Recorded since the run table existed and shown nowhere until this panel. */
  readonly written: number;
  readonly announced: number | null;
  /** `announced - extracted`, or null where the documents state no count. */
  readonly missing: number | null;
}

/**
 * When this source's numbers last moved.
 *
 * Data and not a sentence: the catalog picks the wording and puts the numbers where that
 * language puts them. The comparison itself is the server's, out of a `lag()` window, because
 * every number printed beside a list in this application is counted by the server.
 */
export interface SourceTrend {
  readonly extractedChangedOn: string | null;
  readonly extractedBefore: number | null;
  readonly extractedNow: number | null;
  readonly divergedOn: string | null;
  readonly missing: number | null;
  /** False is not a failure: most sources state no count, and saying so beats a blank. */
  readonly announcedStated: boolean;
}

/**
 * One source, opened: what defines it and what it has been doing.
 *
 * Fetched when a row is opened and never as part of the list, which is refetched on every
 * finished run and every tab focus. `block` and `connection` are file text with the secrets
 * already masked on the server; the browser never sees the unmasked file.
 */
export interface SourceDetail {
  readonly id: string;
  readonly kind: SourceKind;
  readonly enabled: boolean;
  readonly file: SourceFile;
  readonly block: YamlBlock | null;
  /** The connection this source names, because an imap source is half-defined elsewhere. */
  readonly connection: YamlBlock | null;
  readonly runs: readonly SourceRun[];
  readonly trend: SourceTrend;
}

/**
 * Which file defines the source, and where it was read from.
 *
 * `origin` is a path on somebody's machine, which is why it is here and not on the list's
 * envelope: this payload is one deliberate click away, that one is behind every screenshot.
 */
export interface SourceFile {
  readonly name: string;
  readonly layer: ConfigLayer;
  readonly origin: string | null;
}
